docker/service.go

```diff
@@ -186,15 +188,48 @@ func (s *Service) up(imageName string, create bool) error {
 	}
 
 	return s.eachContainer(func(c *Container) error {
-		if outOfSync, err := c.OutOfSync(); err != nil {
-			return err
-		} else if outOfSync {
-			logrus.Warnf("%s needs rebuilding", s.Name())
+		if create {
+			if err := s.recreateIfNeeded(imageName, c); err != nil {
+				return err
+			}
 		}
+
 		return c.Up(imageName)
 	})
 }
 
+func (s *Service) recreateIfNeeded(imageName string, c *Container) error {
+	if s.context.NoRecreate {
+		return nil
+	}
+	outOfSync, err := c.OutOfSync(imageName)
+	if err != nil {
+		return err
+	}
+
+	containerInfo, err := c.findInfo()
+	if containerInfo == nil || err != nil {
+		return err
+	}
+	name := containerInfo.Name[1:]
+
+	logrus.WithFields(logrus.Fields{
+		"outOfSync":     outOfSync,
+		"ForceRecreate": s.context.ForceRecreate,
+		"NoRecreate":    s.context.NoRecreate}).Debug("Going to decide if recreate is needed")
+
+	if s.context.ForceRecreate || outOfSync {
```

Comment:
We could bring that up (before `containerInfo, err := …`) as we already have all information at this point (except the `name` but see comment below) and this would remove one API call (an inspect done in `c.findInfo()`).